### PR TITLE
ENH: Add length attribute to Interval and IntervalIndex

### LINKED ISSUE #18789: ENH: Add a length attribute to Interval and IntervalIndex
#### Problem description
Seems like a pretty standard attribute for an interval, and something that a user may want to know.

Implementation is straight forward: `self.right - self.left`, which could be an argument against adding this, since it's something a user could easily do themselves.   I think the explicit use of `length` would make code more readable.  For example, consider the datetime safe path for `Interval.mid`, the midpoint of an `Interval`:
https://github.com/pandas-dev/pandas/blob/7b4229aa1b95b2050b31cc93e38fda1a72258675/pandas/_libs/interval.pyx#L56-L57

This could be rewritten in a more explicit and clear manner with `Interval.length`:
```python
 # datetime safe version 
 return self.left + 0.5 * self.length
```

Note that I don't think this can be implemented as `__len__` for `Interval`, as I believe `__len__` is required to return an integer, and intervals do not necessarily need to have integer length.

#### Expected Output
Scalar output for an `Interval`:
```python
In [2]: iv = pd.Interval(1, 3)

In [3]: iv
Out[3]: Interval(1, 3, closed='right')

In [4]: iv.length
Out[4]: 2
```

An `Index` with entries denoting the length of each `Interval` for an `IntervalIndex`:
```python
In [5]: idx = pd.IntervalIndex.from_breaks(pd.to_datetime(['20171010', '20171020', '20171111']))

In [6]: idx
Out[6]:
IntervalIndex([(2017-10-10, 2017-10-20], (2017-10-20, 2017-11-11]]
              closed='right',
              dtype='interval[datetime64[ns]]')

In [7]: idx.length
Out[7]: TimedeltaIndex(['10 days', '22 days'], dtype='timedelta64[ns]', freq=None)

```

### PR DESCRIPTION
**Pull request by jschendel**
- [X] closes #18789
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/interval.py [hunk 1]
```diff
@@ -599,6 +599,14 @@ def closed(self):
         """
         return self._closed
 
+    @property
+    def length(self):
+        """
+        Return an Index with entries denoting the length of each Interval in
+        the IntervalIndex
+        """
+        return self.right - self.left
```
Comment (topper-123) [inline pandas/core/indexes/interval.py]: Some intervals don't have a length, if values are categorical-like, e.g. letters (``pd.Interval('b', 'e')``). 

I think maybe the length calculation should be wrapped in a try/except statement, with a helpful error message for when length cannot be calculated (e.g. "not possible to calculate length between 'b' and 'e'" in the above example.

Comment (gfyoung) [reply pandas/core/indexes/interval.py]: @topper-123 : I agree with this assessment.  @jschendel : Do wrap the `length` property as such and add tests.

Comment (jschendel) [reply pandas/core/indexes/interval.py]: Done.  Should probably audit the existing code for more instances of this, as I imagine there are a few places where similar failures occur, e.g. `Interval.mid`.

Comment (gfyoung) [reply pandas/core/indexes/interval.py]: That's probably a good idea.  Depend on how big that change is, you may want to leave that as a follow-up PR though.

Comment (jschendel) [reply pandas/core/indexes/interval.py]: Yeah, planning to do it in a separate PR so I don't delay this one with unrelated work while I look for more instances.  Cleaner than having the instances I immediately see go in here,  and the others in a separate PR too.

**Comment by jreback** [comment]
can you make sure this is in api.rst (though IIRC we are listing these attributes directly in the class).

File: pandas/_libs/interval.pyx [hunk 2]
```diff
@@ -54,7 +54,17 @@ cdef class IntervalMixin(object):
...
+            return self.left + 0.5 * self.length
+
+    @property
+    def length(self):
+        """Return the length of the Interval"""
+        try:
+            return self.right - self.left
+        except TypeError:
+            # length not defined for some types, e.g. string
+            msg = 'cannot compute length between {left} and {right}'
```
Comment (topper-123) [inline pandas/_libs/interval.pyx]: A nitpick perhaps: IMO the form ``msg = 'cannot compute length between {left!r} and {right!r}'`` (note the ``!r``) is better, so variables get outputted as raw strings where relevant in error messages.

In general IMO variables in error messages should use ``!r``, though this is not very common on pandas.

**Comment by jschendel** [comment]
@jreback : added this to api.rst, and made minor changes to the `IntervalIndex` docstring.  Generated the docs and changes look good.

File: doc/source/whatsnew/v0.22.0.txt [hunk 3]
```diff
@@ -141,6 +141,7 @@ Other Enhancements
 - ``IntervalIndex.to_tuples()`` has gained the ``na_tuple`` parameter to control whether NA is returned as a tuple of NA, or NA itself (:issue:`18756`)
 - ``Categorical.rename_categories``, ``CategoricalIndex.rename_categories`` and :attr:`Series.cat.rename_categories`
   can now take a callable as their argument (:issue:`18862`)
+- :class:`Interval` and :class:`IntervalIndex` have gained a ``length`` attribute (:issue:`18789`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: rebase on master. can you move to 0.23 (docs were renamed), prob easiest to just check this file from master and past in new one

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**